include/boost/gil/image_processing/morphology.hpp

```diff
@@ -0,0 +1,193 @@
+                black_hat
+            };
+namespace detail{
+template <typename SrcView, typename DstView, typename Kernel>
+void morph_impl(SrcView const& src_view, DstView const& dst_view, Kernel const& kernel,const int identifier)
+{
+    int flip_ker_row, flip_ker_col, row_boundary, col_boundary;
+    typename channel_type<typename SrcView::value_type>::type max_overlapped_element,min_overlapped_element;
+    for (std::ptrdiff_t view_row = 0; view_row < src_view.height(); ++view_row)
+    {
+        for (std::ptrdiff_t view_col = 0; view_col < src_view.width(); ++view_col)
+        {
+            max_overlapped_element = (std::numeric_limits<typename channel_type<SrcView>::type>::min)();
+            min_overlapped_element = (std::numeric_limits<typename channel_type<SrcView>::type>::max)();
+            for (std::size_t kernel_row = 0; kernel_row < kernel.size(); ++kernel_row)
+            {
+                flip_ker_row = kernel.size() - 1 - kernel_row;      // row index of flipped kernel
+
+                for (std::size_t kernel_col = 0; kernel_col < kernel.size(); ++kernel_col)
+                {
+                    flip_ker_col = kernel.size() - 1 - kernel_col; // column index of flipped kernel
+
+                    // index of input signal, used for checking boundary
+                    row_boundary = view_row + (kernel.center_y() - flip_ker_row);
+                    col_boundary = view_col + (kernel.center_x() - flip_ker_col);
+
+                    // ignore input samples which are out of bound
+                    if (row_boundary >= 0 && row_boundary < src_view.height() &&
+                        col_boundary >= 0 && col_boundary < src_view.width())
+                    {
+                        if(src_view(col_boundary, row_boundary) > max_overlapped_element)
+                            max_overlapped_element = src_view(col_boundary, row_boundary);
+                        if(src_view(col_boundary, row_boundary) < min_overlapped_element)
+                            min_overlapped_element = src_view(col_boundary, row_boundary);
+                    }
+                }
+            }
+            if(identifier)//identifier = 1 for dilation
+                dst_view(view_col, view_row) = max_overlapped_element;
+            else //identifier = 0 for erosion
+                dst_view(view_col, view_row) = min_overlapped_element; 
+        }
+        }
+    }
+
+template <typename SrcView, typename DstView,typename Kernel>
+void morph(SrcView const& src_view, DstView const& dst_view,Kernel const& ker_mat,const int identifier)
+{
+    using namespace boost::gil;
+    BOOST_ASSERT(ker_mat.size() != 0 && src_view.dimensions() == dst_view.dimensions());
+    gil::gray8_image_t intermediate_img(src_view.dimensions());
+    for (std::size_t i = 0; i < src_view.num_channels(); i++)
+    {
+        morph_impl(
+            nth_channel_view(src_view, i),
+            nth_channel_view(view(intermediate_img), i),
```

Comment:
Shouldn't this be:

```
nth_channel_view(view(dst_vuew), i),
```

?

Reply:
I see it now. You need compatible pixel and channel types. Lets just mention in the docs that we require those to be compatible and leave it at that. Let the user deal with that, if there are some common practices, we will just implement them later when they get established. In the worst case, they can write their own adaptor on top of a view.

Reply:
Yes ,my purpose of passing a separate image view was to ensure compatibility,I will implement your advised changes

Reply:
My earlier comment was an attempt to clarify this.

Reply:
@simmplecoder 
> You need compatible pixel and channel types. Lets just mention in the docs that we require those to be compatible

Certainly, a comment/doc is a good idea.
A `static_assert` as a comment/doc is even better, I think :-)

Reply:
@meshtag, please implement the check @mloskot suggested. I believe the right way is to use `gil_function_requires<PixelsCompatibleConcept<SrcPixelType, DstPixelType>>()`.